sql/catalyst/src/main/scala/org/apache/spark/sql/types/YearMonthIntervalType.scala

```diff
@@ -55,16 +65,28 @@ class YearMonthIntervalType private() extends AtomicType {
 
   private[spark] override def asNullable: YearMonthIntervalType = this
 
-  override def typeName: String = "interval year to month"
+  override val typeName: String = (start, end) match {
+    case (YEAR, MONTH) => "interval year to month"
+    case (YEAR, YEAR) => "interval year"
+    case (MONTH, MONTH) => "interval month"
+    case _ => throw new AnalysisException(
+      s"""Invalid interval units: ($start, $end). Supported units:
+         | (0, 0) - interval year
+         | (0, 1) - interval year to month
+         | (1, 1) - interval month
+         |""".stripMargin)
+  }
 }
 
 /**
- * The companion case object and its class is separated so the companion object also subclasses
- * the YearMonthIntervalType class. Otherwise, the companion object would be of type
- * "YearMonthIntervalType$" in byte code. Defined with a private constructor so the companion object
- * is the only possible instantiation.
+ * Extra factory methods and pattern matchers for the year-month interval type.
  *
  * @since 3.2.0
  */
 @Unstable
-case object YearMonthIntervalType extends YearMonthIntervalType
+object YearMonthIntervalType {
```

Comment:
Shall we follow `DecinalType` and extends `AbstractDataType`? This is useful when a function accepts all kinds of year month intervals.

Reply:
I didn't understand this. Could you explain me, please. For now, the code works:
```scala
  test("aaa") {
    def f(y: YearMonthIntervalType): Int = {
      y.defaultSize
    }
    f(YearMonthIntervalType(0, 1))
  }
```
@cloud-fan or mean a different use case?

Reply:
I mean `ExpectsInputTypes.inputTypes`. The "function" I mentioned is SQL function (expression).